libsolidity/interface/CompilerStack.cpp

```diff
@@ -1206,13 +1206,69 @@ Json CompilerStack::ethdebug(Contract const& _contract, bool _runtime) const
 	solAssert(m_stackState >= AnalysisSuccessful, "Analysis was not successful.");
 	solAssert(_contract.contract);
 	solUnimplementedAssert(!isExperimentalSolidity());
-	if (_runtime)
+	Json result = {
+	{"contract",
+		{
+			{"name", _contract.contract->name()},
+			{"definition",
+				{{"sources",
+					{{"id", static_cast<int>(sourceIndices()[_contract.contract->sourceUnitName()])}}
+				}}
+			}}
+		},
+		{"environment ", _runtime ? "call" : "create"},
+		{"instructions", ethdebugInstructions(_contract, _runtime)}
+	};
+	return result;
+}
+
+Json CompilerStack::ethdebugInstructions(Contract const& _contract, bool _runtime) const
+{
+	evmasm::LinkerObject const& object = _runtime ? _contract.runtimeObject : _contract.object;
+	std::shared_ptr<evmasm::Assembly> const& assembly = _runtime ? _contract.evmRuntimeAssembly : _contract.evmAssembly;
+
+	if (assembly)
 	{
-		Json result = Json::object();
-		return result;
+		solUnimplementedAssert(
+			assembly->eofVersion() == std::nullopt && assembly->codeSections().size() == 1,
+			"ethdebug does not yet support EOF.");
 	}
-	Json result = Json::object();
-	return result;
+
+	Json instructions = Json::array();
+	for (size_t i = 0; i < object.bytecodeOffsets.size(); ++i)
+	{
+		size_t offset = object.bytecodeOffsets[i];
+		size_t nextOffset = offset;
+		if (i + 1 < object.bytecodeOffsets.size())
+			nextOffset = object.bytecodeOffsets[i + 1];
+		Json instruction = Json::object();
+		instruction["offset"] = offset;
+		Json operation = Json::object();
+		solAssert(offset < object.bytecode.size());
+		solAssert(nextOffset < object.bytecode.size());
+		operation["mnemonic"] = instructionInfo(static_cast<evmasm::Instruction>(object.bytecode[offset]), m_evmVersion).name;
+		bytes argumentData =
+			bytes(object.bytecode.begin() + static_cast<int>(offset), object.bytecode.begin() + static_cast<int>(nextOffset));
+		if (!argumentData.empty())
+			operation["arguments"] = Json::array({ "0x" + util::toHex(argumentData) });
+		solAssert(assembly->codeSections().at(0).items.size() == object.bytecodeOffsets.size());
+		SourceLocation const& location = assembly->codeSections().at(0).items.at(i).location();
+		Json instructionContext = Json::object();
+		Json source = Json::object();
+		source["id"] = location.sourceName ? static_cast<int>(sourceIndices()[*location.sourceName]) : -1;
+		Json range = Json::object();
+		range["offset"] = location.start;
+		range["length"] = location.end - location.start;
+		Json code = Json::object();
+		code["source"] = source;
+		code["range"] = range;
+		instructionContext["code"] = code;
+		instruction["operation"] = operation;
+		instruction["context"] = instructionContext;
```

Comment:
Could be done in the same style as above :)

```suggestion
		Json const instruction {
			{"offset", offset},
			{"operation", operation},
			{
				"context", 
				{
					{
						"code", 
						{
							{
								"source", 
								{
									{"id", location.sourceName ? static_cast<int>(sourceIndices()[*location.sourceName]) : -1}
								}
							},
							{
								"range", 
								{
									{"offset", location.start},
									{"length", location.end - location.start}
								}
							}
						}
					}
				}
			}
		};
```